src/main/scala/is/hail/expr/types/MatrixType.scala

```diff
@@ -0,0 +1,141 @@
+package is.hail.expr.types
+
+import is.hail.expr.EvalContext
+import is.hail.rvd.OrderedRVType
+import is.hail.variant.{GenomeReference, Genotype}
+
+object MatrixType {
+  def locusType(vType: Type): Type = {
+    vType match {
+      case t: TVariant => TLocus(t.gr)
+      case _ => vType
+    }
+  }
+
+  def apply(
+    globalType: TStruct = TStruct.empty(),
+    sType: Type = TString(),
+    saType: TStruct = TStruct.empty(),
+    vType: Type = TVariant(GenomeReference.defaultReference),
+    vaType: TStruct = TStruct.empty(),
+    genotypeType: TStruct = Genotype.htsGenotypeType): MatrixType = {
+    val lType = locusType(vType)
+
+    val sField = saType.uniqueFieldName("s")
+    val vField = vaType.uniqueFieldName("v")
+    val pkField = vaType.uniqueFieldName("pk")
+    val t = MatrixType(
+      globalType,
+      Array(sField),
+      TStruct(sField -> sType) ++ saType,
+      Array(pkField),
+      Array(pkField, vField),
+      TStruct(pkField -> lType, vField -> vType) ++ vaType,
+      genotypeType)
+    assert(t.sType == sType)
+    assert(t.saType == saType)
+    assert(t.vType == vType)
+    assert(t.vaType == vaType)
+    assert(t.locusType == lType)
+    t
+  }
+}
+
+case class MatrixType(
+  globalType: TStruct,
+  colKey: IndexedSeq[String],
+  colType: TStruct,
+  rowPartitionKey: IndexedSeq[String],
+  rowKey: IndexedSeq[String],
+  // FIXME rowType
+  rowType: TStruct,
+  entryType: TStruct) extends BaseType {
+  assert(colKey.length == 1)
+  val sField: String = colKey(0)
+
+  assert(rowPartitionKey.length == 1)
+  val pkField: String = rowPartitionKey(0)
+
+  assert(rowKey.length == 2)
+  assert(rowKey(0) == pkField)
+  val vField = rowKey(1)
+
+  val vType: Type = rowType.field(vField).typ
+
+  val locusType: Type = MatrixType.locusType(vType)
+
+  val (vaType, _) = rowType.filter(f => f.name != pkField && f.name != vField)
+
+  val sType: Type = colType.field(sField).typ
+
+  val (saType, _) = colType.filter(f => f.name != sField)
+
+  def genotypeType: TStruct = entryType
+
+  // FIXME rename
+  // FIXME needs to be rowType ++ TStruct(entriesField -> TArray(entryType))
+  val rvRowType: TStruct =
+    TStruct(
+      "pk" -> locusType,
```

Comment:
I don't understand why this isn't `rowType.field(pkField).typ`. I understand that Variant is treated differently (should it still be this way?), but if vType != variant, why is the type of pk not `pkField`'s type?

Reply:
No, this was confusing, I deleted kType and pkType.

MatrixType has a single row key, v.  It has a single partition key pk,  which is a Locus or just v.

OrderedRVD supports compound keys so orderedRVType.{kType, pkType} are always structs.  That's the main difference here.